### PR TITLE
add MappedLocationCache (#581)

### PR DESCRIPTION
**Pull request by hbenl**

### CODE REVIEW

File: src/protocol/mapped-location-cache.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+    this.cache.set(cacheKey, mappedLocation);
+    return mappedLocation;
+  }
+
+  /**
+   * Encode a location as a string for use as a cache key
+   */
+  private encodeLocation(location: Location): string {
+    return `${location.scriptId}|${location.line}|${location.column}`;
```
Comment (jasonLaster) [inline src/protocol/mapped-location-cache.ts:52]: I wonder if this could just be `JSON.stringify`

File: src/protocol/mapped-location-cache.ts [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+interface Location {
+  scriptId: string;
+  line: number;
+  column: number;
+}
+
+type MappedLocation = Location[];
+
+export class MappedLocationCache {
```
Comment (jasonLaster) [inline src/protocol/mapped-location-cache.ts:12]: This looks great. I wonder if down the road, we would want to make this a more generic feature of the protocol layer, where you could opt to make certain messages cacheable...

**Review by jasonLaster — COMMENTED** [review]
Looks good to me.

CC @bhackett1024

File: src/protocol/mapped-location-cache.ts (hunk 1 above)
Comment (hbenl) [reply src/protocol/mapped-location-cache.ts:52]: The critical point here is making sure that the string created for 2 locations is the same if `scriptId`, `line` and `column` are the same - I wouldn't trust `JSON.stringify` here: the order of the properties could change and a location could contain additional properties, which we want to ignore here.

**Comment by jasonLaster** [comment]
Screen Shot 2020-09-21 at 6 29 00 PM

**Review by jasonLaster — APPROVED** [review]

**Pull request merged**